# OHE

In [ ]:
corpus = [
    "the cat sat on the mat",
    "the dog sat on the log",
    "cats and dogs are great pets",
]

# cat, sat, mat, dog, pets, great
# and, are, the, the --> too common

In [4]:
import pandas as pd

In [5]:
words = ["cat", "dog", "cat", "bird"]
df_words = pd.DataFrame({"word": words})

one_hot = pd.get_dummies(df_words["word"]).astype(int)
print(one_hot)

   bird  cat  dog
0     0    1    0
1     0    0    1
2     0    1    0
3     1    0    0


In [6]:
# # Problem with OHE
# - 1. One Vector Per Word, not per document
# - 2. No notion of frequency
# - 3. Massive, sparse vectors at real world vocabulary sizes

# BAG OF WORDS

In [7]:
# Bag of words directly addresses the problem : No notion of frequency that OHE has

corpus = [
    "the cat sat on the mat",
    "the dog sat on the log",
    "cats and dogs are great pets",
]

# corpus[0] --> []

# "dog bites man"
# "man bites dog"
# -> above will generate same BOW(BAG of WORDS) vector
# mat, log, pets, great

# CORPUS LEVEL

In [8]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
bow_matrix = vectorizer.fit_transform(corpus)

vocab = vectorizer.get_feature_names_out()
print("Vocabulary:", vocab)

bow_df = pd.DataFrame(bow_matrix.toarray(), columns=vocab)
bow_df

Vocabulary: ['and' 'are' 'cat' 'cats' 'dog' 'dogs' 'great' 'log' 'mat' 'on' 'pets'
 'sat' 'the']


,and,are,cat,cats,dog,dogs,great,log,mat,on,pets,sat,the
0,0,0,1,0,0,0,0,0,1,1,0,1,2
1,0,0,0,0,1,0,0,1,0,1,0,1,2
2,1,1,0,1,0,1,1,0,0,0,1,0,0


In [ ]:
corpus = [
    "the cat sat on the mat",
    "the dog sat on the log",
    "cats and dogs are great pets",
    "dog bites man",
    "man bites dog",
    "man can be a dog"
]


problems :
1. cat / cats have two seperate columns --> solution : lemma
2. BOW throws away word order -->
Solution for 2.

We cannot sort the columns alphabetically, but what we can do is, keep the words together and count them

"The", "cat" --> "The cat"

"cat", "sat" --> "cat sat"

In [ ]:
bigram_vectorizer = CountVectorizer(ngram_range=(2, 2))
bigram_matrix = bigram_vectorizer.fit_transform(corpus)

bigram_vocab = bigram_vectorizer.get_feature_names_out()
print("Bigram vocabulary:", bigram_vocab)

pd.DataFrame(bigram_matrix.toarray(), columns=bigram_vocab)

Bigram vocabulary: ['and dogs' 'are great' 'be dog' 'bites dog' 'bites man' 'can be'
 'cat sat' 'cats and' 'dog bites' 'dog sat' 'dogs are' 'great pets'
 'man bites' 'man can' 'on the' 'sat on' 'the cat' 'the dog' 'the log'
 'the mat']


,and dogs,are great,be dog,bites dog,bites man,can be,cat sat,cats and,dog bites,dog sat,dogs are,great pets,man bites,man can,on the,sat on,the cat,the dog,the log,the mat
0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,1,1,1,0,0,1
1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,1,1,0,1,1,0
2,1,1,0,0,0,0,0,1,0,0,1,1,0,0,0,0,0,0,0,0
3,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0
4,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0
5,0,0,1,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,0,0


In [ ]:
# include both unigram and bigram
bigram_vectorizer = CountVectorizer(ngram_range=(1, 2))
bigram_matrix = bigram_vectorizer.fit_transform(corpus)

bigram_vocab = bigram_vectorizer.get_feature_names_out()
print("Bigram vocabulary:", bigram_vocab)

pd.DataFrame(bigram_matrix.toarray(), columns=bigram_vocab)

Bigram vocabulary: ['and' 'and dogs' 'are' 'are great' 'be' 'be dog' 'bites' 'bites dog'
 'bites man' 'can' 'can be' 'cat' 'cat sat' 'cats' 'cats and' 'dog'
 'dog bites' 'dog sat' 'dogs' 'dogs are' 'great' 'great pets' 'log' 'man'
 'man bites' 'man can' 'mat' 'on' 'on the' 'pets' 'sat' 'sat on' 'the'
 'the cat' 'the dog' 'the log' 'the mat']


,and,and dogs,are,are great,be,be dog,bites,bites dog,bites man,can,...,on,on the,pets,sat,sat on,the,the cat,the dog,the log,the mat
0,0,0,0,0,0,0,0,0,0,0,...,1,1,0,1,1,2,1,0,0,1
1,0,0,0,0,0,0,0,0,0,0,...,1,1,0,1,1,2,0,1,1,0
2,1,1,1,1,0,0,0,0,0,0,...,0,0,1,0,0,0,0,0,0,0
3,0,0,0,0,0,0,1,0,1,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,1,1,0,0,...,0,0,0,0,0,0,0,0,0,0
5,0,0,0,0,1,1,0,0,0,1,...,0,0,0,0,0,0,0,0,0,0


In [ ]:
# trigram only
trigram_vectorizer = CountVectorizer(ngram_range=(3, 3))
trigram_matrix = trigram_vectorizer.fit_transform(corpus)

trigram_vocab = trigram_vectorizer.get_feature_names_out()
print("Trigram vocabulary:", trigram_vocab)

pd.DataFrame(trigram_matrix.toarray(), columns=trigram_vocab)

problem with BOW/ngrams :
1. every word counts equally

ngrams(1,2) --> the --> 2 --> remove this with stop word --> fix the issue

corpus = [
    "the cat sat on the mat",
    "the dog sat on the log",
    "cats and dogs are great pets",
]

cat, sat, mat, dog, pets, great
and, are, the, the --> too common

# TF-IDF

- TF - how often a word appreas in this document  (Term frequency)
- IDF - how rare that word is across the whole corpus (Inverse document frequency)
- Score/value for a word = TF * IDF

In [ ]:
corpus = [
    "the cat sat on the mat",
    "the dog sat on the log",
    "cats and dogs are great pets",
    "the the the the the the the the"
]

In [ ]:
[d.split() for d in corpus]

[['the', 'cat', 'sat', 'on', 'the', 'mat'],
 ['the', 'dog', 'sat', 'on', 'the', 'log'],
 ['cats', 'and', 'dogs', 'are', 'great', 'pets'],
 ['the', 'the', 'the', 'the', 'the', 'the', 'the', 'the']]

In [ ]:
import math

In [ ]:
N = len(corpus)
print(N)

docs = [d.split() for d in corpus]

def term_freq(doc, term):
    return doc.count(term)

def doc_freq(term):
    return sum(1 for d in docs if term in d)

def idf(term):
    df = doc_freq(term)
    return math.log((1+N)/(1+df))+1

def tfidf(term, doc):
    tf = term_freq(doc, term)
    return tf * idf(term)

4


In [ ]:
final_dict = {}
for i, document in enumerate(corpus):
    final_dict[i] = {}
    for word in document.split():
        word_tfidf = tfidf(word, document)
        final_dict[i][word] = word_tfidf

print(final_dict)

{0: {'the': 2.4462871026284194, 'cat': 1.916290731874155, 'sat': 1.5108256237659907, 'on': 1.5108256237659907, 'mat': 1.916290731874155}, 1: {'the': 2.4462871026284194, 'dog': 1.916290731874155, 'sat': 1.5108256237659907, 'on': 1.5108256237659907, 'log': 1.916290731874155}, 2: {'cats': 1.916290731874155, 'and': 1.916290731874155, 'dogs': 1.916290731874155, 'are': 1.916290731874155, 'great': 1.916290731874155, 'pets': 1.916290731874155}, 3: {'the': 9.785148410513678}}


{0: {'the': 2.4462871026284194, 'cat': 1.916290731874155, 'sat': 1.5108256237659907, 'on': 1.5108256237659907, 'mat': 1.916290731874155},

1: {'the': 2.4462871026284194, 'dog': 1.916290731874155, 'sat': 1.5108256237659907, 'on': 1.5108256237659907, 'log': 1.916290731874155},

2: {'cats': 1.916290731874155, 'and': 1.916290731874155, 'dogs': 1.916290731874155, 'are': 1.916290731874155, 'great': 1.916290731874155, 'pets': 1.916290731874155},

3: {'the': 9.785148410513678}}


In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd

tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(corpus)

vocab = tfidf_vectorizer.get_feature_names_out()
tfidf_df = pd.DataFrame(tfidf_matrix.toarray(), columns=vocab).round(3)
tfidf_df

,and,are,cat,cats,dog,dogs,great,log,mat,on,pets,sat,the
0,0.000,0.000,0.453,0.000,0.000,0.000,0.000,0.000,0.453,0.357,0.000,0.357,0.578
1,0.000,0.000,0.000,0.000,0.453,0.000,0.000,0.453,0.000,0.357,0.000,0.357,0.578
2,0.408,0.408,0.000,0.408,0.000,0.408,0.408,0.000,0.000,0.000,0.408,0.000,0.000
3,0.000,0.000,0.000,0.000,0.000,0.000,0.000,0.000,0.000,0.000,0.000,0.000,1.000
